pallets/moonbeam-foreign-assets/src/evm.rs

```diff
@@ -0,0 +1,371 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{AssetId, Error, Pallet};
+use ethereum_types::{BigEndianHash, H160, H256, U256};
+use fp_evm::{ExitReason, ExitSucceed};
+use frame_support::ensure;
+use frame_support::pallet_prelude::Weight;
+use pallet_evm::{GasWeightMapping, Runner};
+use precompile_utils::prelude::*;
+use precompile_utils::solidity::codec::{Address, BoundedString};
+use precompile_utils::solidity::Codec;
+use precompile_utils_macro::keccak256;
+use sp_runtime::traits::ConstU32;
+use sp_runtime::DispatchError;
+use sp_std::vec::Vec;
+use xcm::latest::Error as XcmError;
+
+const ERC20_CALL_MAX_CALLDATA_SIZE: usize = 4 + 32 + 32; // selector + address + uint256
+const ERC20_CREATE_MAX_CALLDATA_SIZE: usize = 16 * 1024; // 16Ko
+
+// Hardcoded gas limits (from manueal binary search)
+const ERC20_CREATE_GAS_LIMIT: u64 = 3_367_000; // highest failure: 3_366_000
+const ERC20_BURN_FROM_GAS_LIMIT: u64 = 155_000; // highest failure: 154_000
+const ERC20_MINT_INTO_GAS_LIMIT: u64 = 155_000; // highest failure: 154_000
+const ERC20_PAUSE_GAS_LIMIT: u64 = 150_000; // highest failure: 149_500
+const ERC20_TRANSFER_GAS_LIMIT: u64 = 155_000; // highest failure: 154_000
+const ERC20_UNPAUSE_GAS_LIMIT: u64 = 150_000; // highest failure: 149_500
+
+pub enum EvmError {
+	BurnFromFail,
+	ContractReturnInvalidValue,
+	DispatchError(DispatchError),
+	EvmCallFail,
+	MintIntoFail,
+	TransferFail,
+}
+
+impl From<DispatchError> for EvmError {
+	fn from(e: DispatchError) -> Self {
+		Self::DispatchError(e)
+	}
+}
+
+impl From<EvmError> for XcmError {
+	fn from(error: EvmError) -> XcmError {
+		match error {
+			EvmError::BurnFromFail => {
+				XcmError::FailedToTransactAsset("Erc20 contract call burnFrom fail")
+			}
+			EvmError::ContractReturnInvalidValue => {
+				XcmError::FailedToTransactAsset("Erc20 contract return invalid value")
+			}
+			EvmError::DispatchError(err) => {
+				log::debug!("dispatch error: {:?}", err);
+				Self::FailedToTransactAsset("storage layer error")
+			}
+			EvmError::EvmCallFail => XcmError::FailedToTransactAsset("Fail to call erc20 contract"),
+			EvmError::MintIntoFail => {
+				XcmError::FailedToTransactAsset("Erc20 contract call mintInto fail")
+			}
+			EvmError::TransferFail => {
+				XcmError::FailedToTransactAsset("Erc20 contract call transfer fail")
+			}
+		}
+	}
+}
+
+#[derive(Codec)]
+#[cfg_attr(test, derive(Debug))]
+struct ForeignErc20ConstructorArgs {
+	owner: Address,
+	decimals: u8,
+	ticker: BoundedString<ConstU32<64>>,
```

Comment:
Symbol could be more representative here 
```suggestion
	symbol: BoundedString<ConstU32<64>>,
```